pkg/sensors/tracing/loader.go

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Tetragon
+
+package tracing
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"reflect"
+	"syscall"
+	"unsafe"
+
+	"github.com/cilium/tetragon/pkg/api/ops"
+	"github.com/cilium/tetragon/pkg/api/tracingapi"
+	"github.com/cilium/tetragon/pkg/bpf"
+	"github.com/cilium/tetragon/pkg/grpc/tracing"
+	"github.com/cilium/tetragon/pkg/k8s/apis/cilium.io/v1alpha1"
+	"github.com/cilium/tetragon/pkg/kernels"
+	"github.com/cilium/tetragon/pkg/logger"
+	"github.com/cilium/tetragon/pkg/observer"
+	"github.com/cilium/tetragon/pkg/sensors"
+	"github.com/cilium/tetragon/pkg/sensors/program"
+	"golang.org/x/sys/unix"
+)
+
+var (
+	loader = program.Builder(
+		"bpf_loader.o",
+		"perf_event_mmap_output",
+		"kprobe/perf_event_mmap_output",
+		"loader_kprobe",
+		"loader",
+	)
+
+	idsMap = program.MapBuilder("ids_map", loader)
+
+	loaderEnabled bool
+
+	loaderFds []int
+)
+
+type loaderSensor struct {
+	name string
+}
+
+func init() {
+	loader := &loaderSensor{
+		name: "loader sensor",
+	}
+	sensors.RegisterProbeType("loader", loader)
+	sensors.RegisterTracingSensorsAtInit(loader.name, loader)
+
+	observer.RegisterEventHandlerAtInit(ops.MSG_OP_LOADER, handleLoader)
+}
+
+func GetLoaderSensor() *sensors.Sensor {
+	return &sensors.Sensor{
+		Name:  "__loader__",
+		Progs: []*program.Program{loader},
+		Maps:  []*program.Map{idsMap},
+	}
+}
+
+func hasLoaderEvents() bool {
+	return bpf.HasBuildId() && kernels.MinKernelVersion("5.19.0")
+}
+
+func (k *loaderSensor) SpecHandler(raw interface{}) (*sensors.Sensor, error) {
+	spec, ok := raw.(*v1alpha1.TracingPolicySpec)
+	if !ok {
+		s, ok := reflect.Indirect(reflect.ValueOf(raw)).FieldByName("TracingPolicySpec").Interface().(v1alpha1.TracingPolicySpec)
+		if !ok {
+			return nil, nil
+		}
+		spec = &s
+	}
+	if spec.Loader {
+		if !hasLoaderEvents() {
+			return nil, fmt.Errorf("Loader event are not supported on running kernel")
+		}
+		loaderEnabled = true
+		return GetLoaderSensor(), nil
+	}
+	return nil, nil
+}
+
+func createLoaderEvents() error {
+	attr := &unix.PerfEventAttr{
+		Type:        unix.PERF_TYPE_SOFTWARE,
+		Config:      unix.PERF_COUNT_SW_BPF_OUTPUT,
+		Sample_type: unix.PERF_SAMPLE_RAW,
+		Bits:        unix.PerfBitMmap | unix.PerfBitMmap2 | bpf.PerfBitBuildId,
+	}
+
+	nCpus := bpf.GetNumPossibleCPUs()
+
+	// We create perf event for each cpu and uploade their ids
+	// into 'ids_map' for the bpf program to recognize our event.
+
+	var ids []uint64
+
+	for cpu := 0; cpu < nCpus; cpu++ {
+		fd, err := unix.PerfEventOpen(attr, -1, cpu, -1, unix.PERF_FLAG_FD_CLOEXEC)
+		if err != nil {
+			return fmt.Errorf("can't create perf event: %w", err)
+		}
+		loaderFds = append(loaderFds, fd)
```

Comment:
What is the purpose of `loaderFds`? AFAICT it is not used somewhere else. Was the intention to close the files at some point?

Reply:
I need fds in loaderFds to stay alive after leaving the function,
but I might be too paranoid in here, I'll double check and most likely change this,
will add more comments